ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014-2017 Marc de Verdelhan, 2017-2019 Ta4j Organization & respective
+ * authors (see AUTHORS)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core.trading.rules;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.ta4j.core.BarSeries;
+import org.ta4j.core.BaseBarSeries;
+import org.ta4j.core.Indicator;
+import org.ta4j.core.indicators.helpers.FixedDecimalIndicator;
+import org.ta4j.core.num.Num;
+import org.ta4j.core.trading.rules.helper.ChainLink;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class ChainRuleTest {
+
+    private ChainRule chainRule;
+
+    @Before
+    public void setUp() {
+        BarSeries series = new BaseBarSeries();
+        Indicator<Num> indicator = new FixedDecimalIndicator(series, 6, 5, 8, 5, 1, 10, 2, 30);
+        UnderIndicatorRule underIndicatorRule = new UnderIndicatorRule(indicator, series.numOf(5));
+        OverIndicatorRule overIndicatorRule = new OverIndicatorRule(indicator, 7);
+        IsEqualRule isEqualRule = new IsEqualRule(indicator, 5);
+        chainRule = new ChainRule(underIndicatorRule, new ChainLink(overIndicatorRule, 3), new ChainLink(isEqualRule, 2));
+    }
+
+    @Test
+    public void isSatisfied() {
+        assertFalse(chainRule.isSatisfied(0));
+        assertTrue(chainRule.isSatisfied(4));
+        assertTrue(chainRule.isSatisfied(6));
```

Comment:
why should ChainRule return true for index `6` ? 
   
index 6 is 2

6, 5, 8, 5, 1, 10, **2**

- initial rule:         **2** < 5 true
- overIndicator (3):    **10**  > 7  || **1**  > 7 || **5** > 7 true 
- isEqualRule (2):       **2** == 5 || **10** == 5 **false**

Reply:
Hi @team172011 

there is a missunderstanding how this code works:

- initial rule: 2 < 5 true: This is an easy one, the rule is satisfied immediately and the next rule in the chain starts with the **same** index
- overIndicator (3): is therefore true at index 5, next rule starts on index 5
- isEqualRule (2): is therefore true on index 3 ;)

Reply:
okay, this way this rule makes more sense :)